File: packages/asdl-slots/src/asdl_slots/lifecycle/release_cleanup.py
"""Cleanup planning and execution for released slot targets."""

from __future__ import annotations

from collections.abc import Sequence

from asdl_core.gh.types import PRGatewayFailure, PRLookupMiss
from asdl_slots.context import SlotsCliContext
from asdl_slots.lifecycle.outcomes import (
    FreedSlot,
    SlotFreeCleanupAction,
    SlotFreeCleanupResult,
    SlotFreeCleanupStatus,
)

SLOT_RELEASE_ALL_CLEANUP_ACTIONS: tuple[SlotFreeCleanupAction, ...] = (
    "pr",
    "local_branch",
)


def plan_release_cleanup(
    slots_ctx: SlotsCliContext,
    targets: Sequence[FreedSlot],
    cleanup_actions: Sequence[SlotFreeCleanupAction],
    *,
    trunk_branch: str | None = None,
) -> tuple[SlotFreeCleanupResult, ...]:
    """Plan cleanup entries for release targets without mutating PRs or branches."""
    return _cleanup_for_targets(
        slots_ctx,
        targets,
        cleanup_actions,
        trunk_branch=trunk_branch,
        execute=False,
    )


def execute_release_cleanup(
    slots_ctx: SlotsCliContext,
    targets: Sequence[FreedSlot],
    cleanup_actions: Sequence[SlotFreeCleanupAction],
    *,
    trunk_branch: str | None = None,
) -> tuple[SlotFreeCleanupResult, ...]:
    """Run requested cleanup actions for released slots."""
    return _cleanup_for_targets(
        slots_ctx,
        targets,
        cleanup_actions,
        trunk_branch=trunk_branch,
        execute=True,
    )


def _cleanup_for_targets(
    slots_ctx: SlotsCliContext,
    targets: Sequence[FreedSlot],
    cleanup_actions: Sequence[SlotFreeCleanupAction],
    *,
    trunk_branch: str | None,
    execute: bool,
) -> tuple[SlotFreeCleanupResult, ...]:
    if not targets or not cleanup_actions:
        return ()

    needs_trunk = any(action == "local_branch" for action in cleanup_actions)
    resolved_trunk = trunk_branch
    if needs_trunk and resolved_trunk is None:
        resolved_trunk = slots_ctx.git.get_trunk_branch()

    results: list[SlotFreeCleanupResult] = []
    for target in targets:
        for action in cleanup_actions:
            if action == "pr":
                result = _cleanup_pr(slots_ctx, target, execute=execute)
            elif action == "local_branch":
                assert resolved_trunk is not None
                result = _cleanup_local_branch(
                    slots_ctx,
                    target,
                    trunk_branch=resolved_trunk,
                    execute=execute,
                )
            else:
                raise ValueError(f"unknown cleanup action: {action}")
            results.append(result)
            if result.status == "error":
                return tuple(results)
    return tuple(results)


def _cleanup_pr(
    slots_ctx: SlotsCliContext,
    target: FreedSlot,
    *,
    execute: bool,
) -> SlotFreeCleanupResult:
    pr_result = slots_ctx.pr.get_pr_for_branch(target.branch_name)
    if isinstance(pr_result, PRLookupMiss):
        return _cleanup_result(
            target,
            "pr",
            "skipped",
            message="no matching PR",
        )
    if isinstance(pr_result, PRGatewayFailure):
        return _cleanup_result(
            target,
            "pr",
            "error",
            message=_pr_failure_message(pr_result),
        )

    if pr_result.state in ("CLOSED", "MERGED"):
        return _cleanup_result(
            target,
            "pr",
            "skipped",
            pr_number=pr_result.number,
            message=f"PR is already {pr_result.state.lower()}",
        )

    if not execute:
        return _cleanup_result(target, "pr", "planned", pr_number=pr_result.number)

    close_result = slots_ctx.pr.close_pr(pr_result.number)
    if isinstance(close_result, PRGatewayFailure):
        return _cleanup_result(
            target,
            "pr",
            "error",
            pr_number=pr_result.number,
            message=_pr_failure_message(close_result),
        )
    return _cleanup_result(target, "pr", "success", pr_number=pr_result.number)


def _cleanup_local_branch(
    slots_ctx: SlotsCliContext,
    target: FreedSlot,
    *,
    trunk_branch: str,
    execute: bool,
) -> SlotFreeCleanupResult:
    if target.branch_name == trunk_branch:
        return _cleanup_result(
            target,
            "local_branch",
            "error",
            message=f"refusing to delete trunk branch {trunk_branch}",
        )
    if not execute:
        return _cleanup_result(target, "local_branch", "planned")

    if not slots_ctx.git.branch_exists(target.branch_name):
        return _cleanup_result(target, "local_branch", "skipped", message="already absent")

    failure = slots_ctx.git.delete_local_branch(target.branch_name, force=True)
    if failure is not None:
        if _is_missing_local_branch_failure(failure.message, target.branch_name):
            return _cleanup_result(target, "local_branch", "skipped", message="already absent")
        return _cleanup_result(target, "local_branch", "error", message=failure.message)
    return _cleanup_result(target, "local_branch", "success")


def _is_missing_local_branch_failure(message: str, branch: str) -> bool:
    lowered = message.lower()
    quoted = f"branch '{branch.lower()}' not found"
    return quoted in lowered


def _cleanup_result(
    target: FreedSlot,
    action: SlotFreeCleanupAction,
    status: SlotFreeCleanupStatus,
    *,
    pr_number: int | None = None,
    message: str | None = None,
) -> SlotFreeCleanupResult:
    return SlotFreeCleanupResult(
        slot_name=target.slot_name,
        branch_name=target.branch_name,
        action=action,
        status=status,
        pr_number=pr_number,
        message=message,
    )


def _pr_failure_message(failure: PRGatewayFailure) -> str:
    return failure.stderr or failure.stdout or f"gh exited {failure.returncode}"

File: ts/packages/slot/src/lifecycle/release-cleanup.ts
import type { RepoSlotContext } from "../context.ts";
import { prFailureMessage } from "../gateways/pr.ts";
import type { FreedSlot } from "./release-target.ts";

export type SlotFreeCleanupAction = "pr" | "local_branch";
export type SlotFreeCleanupStatus = "planned" | "success" | "skipped" | "error";

export interface SlotFreeCleanupResult {
	slot_name: string;
	branch_name: string;
	action: SlotFreeCleanupAction;
	status: SlotFreeCleanupStatus;
	pr_number: number | null;
	message: string | null;
}

export const SLOT_RELEASE_ALL_CLEANUP_ACTIONS = ["pr", "local_branch"] as const satisfies readonly SlotFreeCleanupAction[];

export interface PlanReleaseCleanupOptions {
	ctx: RepoSlotContext;
	targets: readonly FreedSlot[];
	cleanupActions: readonly SlotFreeCleanupAction[];
	trunkBranch?: string | undefined;
}

export interface ExecuteReleaseCleanupOptions {
	ctx: RepoSlotContext;
	targets: readonly FreedSlot[];
	cleanupActions: readonly SlotFreeCleanupAction[];
	trunkBranch?: string | undefined;
}

interface CleanupForTargetsOptions {
	ctx: RepoSlotContext;
	targets: readonly FreedSlot[];
	cleanupActions: readonly SlotFreeCleanupAction[];
	trunkBranch?: string | undefined;
	shouldExecute: boolean;
}

interface CleanupLocalBranchOptions {
	ctx: RepoSlotContext;
	target: FreedSlot;
	trunkBranch: string;
	shouldExecute: boolean;
}

interface CleanupResultOptions {
	target: FreedSlot;
	action: SlotFreeCleanupAction;
	status: SlotFreeCleanupStatus;
	prNumber?: number | undefined;
	message?: string | undefined;
}

export async function planReleaseCleanup(options: PlanReleaseCleanupOptions): Promise<readonly SlotFreeCleanupResult[]> {
	return await cleanupForTargets({ ...options, shouldExecute: false });
}

export async function executeReleaseCleanup(options: ExecuteReleaseCleanupOptions): Promise<readonly SlotFreeCleanupResult[]> {
	return await cleanupForTargets({ ...options, shouldExecute: true });
}

async function cleanupForTargets(options: CleanupForTargetsOptions): Promise<readonly SlotFreeCleanupResult[]> {
	if (options.targets.length === 0 || options.cleanupActions.length === 0) return [];
	const needsTrunk = options.cleanupActions.includes("local_branch");
	const trunkBranch = needsTrunk ? options.trunkBranch ?? await options.ctx.git.getTrunkBranch() : null;
	const results: SlotFreeCleanupResult[] = [];
	for (const target of options.targets) {
		for (const action of options.cleanupActions) {
			const result = action === "pr" ? await cleanupPr(options.ctx, target, options.shouldExecute) : await cleanupLocalBranch({ ctx: options.ctx, target, trunkBranch: requireTrunkBranch(trunkBranch), shouldExecute: options.shouldExecute });
			results.push(result);
			if (result.status === "error") return results;
		}
	}
	return results;
}

async function cleanupPr(ctx: RepoSlotContext, target: FreedSlot, shouldExecute: boolean): Promise<SlotFreeCleanupResult> {
	const lookup = await ctx.pr.getPrForBranch(target.branch_name);
	if (lookup.type === "miss") return cleanupResult({ target, action: "pr", status: "skipped", message: "no matching PR" });
	if (lookup.type === "failure") return cleanupResult({ target, action: "pr", status: "error", message: prFailureMessage(lookup.failure) });
	if (lookup.pr.state === "CLOSED" || lookup.pr.state === "MERGED") return cleanupResult({ target, action: "pr", status: "skipped", prNumber: lookup.pr.number, message: `PR is already ${lookup.pr.state.toLowerCase()}` });
	if (!shouldExecute) return cleanupResult({ target, action: "pr", status: "planned", prNumber: lookup.pr.number });
	const close = await ctx.pr.closePr(lookup.pr.number);
	if (close.type === "failure") return cleanupResult({ target, action: "pr", status: "error", prNumber: lookup.pr.number, message: prFailureMessage(close.failure) });
	return cleanupResult({ target, action: "pr", status: "success", prNumber: lookup.pr.number });
}

async function cleanupLocalBranch(options: CleanupLocalBranchOptions): Promise<SlotFreeCleanupResult> {
	if (options.target.branch_name === options.trunkBranch) return cleanupResult({ target: options.target, action: "local_branch", status: "error", message: `refusing to delete trunk branch ${options.trunkBranch}` });
	if (!options.shouldExecute) return cleanupResult({ target: options.target, action: "local_branch", status: "planned" });
	if (!(await options.ctx.git.branchExists(options.target.branch_name))) return cleanupResult({ target: options.target, action: "local_branch", status: "skipped", message: "already absent" });
	const failure = await options.ctx.git.deleteLocalBranch(options.target.branch_name, { shouldForce: true });
	if (failure === null) return cleanupResult({ target: options.target, action: "local_branch", status: "success" });
	if (isMissingLocalBranchFailure(failure.message, options.target.branch_name)) return cleanupResult({ target: options.target, action: "local_branch", status: "skipped", message: "already absent" });
	return cleanupResult({ target: options.target, action: "local_branch", status: "error", message: failure.message });
}

function cleanupResult(options: CleanupResultOptions): SlotFreeCleanupResult {
	return { slot_name: options.target.slot_name, branch_name: options.target.branch_name, action: options.action, status: options.status, pr_number: options.prNumber ?? null, message: options.message ?? null };
}

function requireTrunkBranch(trunkBranch: string | null): string {
	if (trunkBranch === null) throw new Error("local branch cleanup requires a resolved trunk branch");
	return trunkBranch;
}

function isMissingLocalBranchFailure(message: string, branch: string): boolean {
	return message.toLowerCase().includes(`branch '${branch.toLowerCase()}' not found`);
}
